common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -214,168 +107,92 @@ public void run() {
     public Void processChallenge(@NonNull final ClientCertRequest request) {
         //final String methodTag = TAG + ":processChallenge";
         //If a smartcard device is connected, proceed with smartcard CBA.
-        synchronized (sDeviceLock) {
-            if (mDevice != null) {
-                handleSmartcardCertAuth(request);
-                return null;
-            }
+        if (mSmartcardCertBasedAuthManager.isDeviceConnected()) {
```

Comment:
I'm not sure about providing a challenge handler to `AzureActiveDirectoryWebViewClient`'s constructor because at the time of `AzureActiveDirectoryWebViewClient`'s instantiation, we don't necessarily know if we're going with smartcard CBA or on-device yet. We have to check right when we receive the `ClientCertRequest`. 
If I write out a way this could work (for my understanding):
-  `mSmartcardCertBasedAuthManager` should be injected into `AzureActiveDirectoryWebViewClient`, and `AzureActiveDirectoryWebViewClient` will need to take care of starting usb discovery (instead of the challenge handler). 
- Factory takes in `mSmartcardCertBasedAuthManager` as a parameter for its new method (we'lll call it `isDeviceConnected`) and uses it to determine whether a smartcard is connected, then returns a challenge handler based on that (smartcard CBA challenge handler will need the manager to be injected).

Reply:
- You could probably instantiate an IChallengeHandler object in onReceivedClientCertRequest()
- When that happens, then you can 

``` Java

final ISmartcardCertBasedAuthManager mYubikitManager;
final ISmartcardCertBasedAuthManager mSomeOtherManager;
...

// Initializes all the manager. give it time to do discovery.
void initialize(){
      mYubikitManager = new YubiKitCertBasedAuthManager(context);
      mYubikitManager.startDiscovery();
      
      mSomeOtherManager = ..... some other smartcard provider.
      mSomeOtherManager.startDiscovery();
}

IChallengeHandler getChallengeHandler(final Context context) {
        if (mYubikitManager.isDeviceConnected()){
              return new SmartcardCertBasedAuthChallengeHandler(mYubiKitManager);
        }
        
        if (mSomeOtherManager.isDeviceConnected()){
              return new OnDeviceCertBasedAuthChallengeHandler(mSomeOtherManager);
        }
        
        return new OnDeviceCertAuthManager();
}

```

Reply:
`AzureActiveDirectoryWebViewClient` now has an instance of `ClientCertAuthFactory` that holds the manager and dialogholder. I ended up injecting the manager and dialogholder into the constructor. In the future, I think the manager parameter can turn into a `List<AbstractSmartcardCertBasedAuthManager>`, where the list contains a manager per smartcard that we support.